### ISSUE TITLE
Review code for LinearProjectionRB

### ISSUE THREAD

**Issue by mpuerrer** [dingo-gw/dingo#2]
I'm looking at the commit with git hash f1445effb1012ad0073479cf3b2add0724325f00

This looks pretty nice, I have some comments and suggestions below.

## LinearProjectionRB

  * remove unused imports:
 
 ```
    import numpy as np
    from torch.nn import functional as F
    from nflows.nn.nets.resnet import ResidualBlock
 ```
  
  * main docstring:
    - "Each block has num_channels >= 2 channels. Channel 0 and 1 represent the real and imaginary part, respectively."
      - What happens if num_channels > 2? What is the interpretation? The code in init_layers assumes num_channels = 2.
    - Give an interpretation for "num_bins": i.e. time of frequency bins of a vector.
    - "the output dimension is 2 * n_rb * num_blocks"
      - Put that after the Parameter description. Again there is an assumption that
        num_channels = 2 in "2 * n_rb * num_blocks", i.e. complex input vectors [Re, Im]^T in the 
        flattened channels and bins dimensions.
    - "V_rb : list of np.arrays      list with v matrices of the reduced basis SVD projection"
      - Say what a "v" matrix is. It assumes an SVD decomposition of a matrix into U @ S @ V^h
    - Say what the dimensionality of the compressed output is.
      
  * __init__():
    * Maybe put line "self.num_blocks, self.num_channels, self.num_bins = self.input_dims" before
    call to "self.test_dimensions()"? Then could refer to the self.num_blocks etc in the asserts
    to make it easier to understand.
    - "self.n_rb * 2": again assuming num_channels = 2

  * output_dim(): again assuming num_channels = 2

  * test_dimensions():
    - Consider using self.num_blocks, etc as suggested above for clarity.
    - I would argue that raising exceptions would be better than using asserts
    for checking user input. E.g.

 ```
      if len(input_dims) != 3:
          raise ValueError("Exactly 3 axes required: blocks, channels, bins")
 ```

    - See e.g. https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/ for a discussion
    of when to use assertions vs exceptions.
  
  * init_layers():
    - "# loop through layers and initialize them individually" -> turn this comment into a doc string!
    - Say in docstring what shape and its elements V_rb have
    - Perhaps use capital letters for matrices, i.e. "v" -> "V"?
    - Perhaps use shorthands for improved readability:

 ```
        k = self.n_rb
        n = self.num_bins
        w[:k, :n] = ...
 ```

    - Use `.transpose` instead of `permute(1, 0)` since there are only two indices involved
    
  * forward():
    - Perhaps use a list comprehension.    
    - `x = torch.cat(out, dim=1)` I wouldn't call the output x when the input is called x.
      A new name would just be a reference, right?

## testutils

  * generate_1d_datasets_and_reduced_basis()
    - The code in plot would fail if plot=True, since n_rb is not defined.
    
  * project_onto_reduced_basis_and_concat():
    - break long lines:

```
        projections = [[(y[:] @ V_rb[:, :n_rb]).real, (y[:] @ V_rb[:, :n_rb]).imag]
            for y, V_rb in zip(y_list, V_rb_list)]
 ```

    - list comprehensions with multiple iterations are hard to read, because the order the
    loops is kind of backward.

 ```
        projections = [item for sublist in projections for item in sublist]
 ```
 
      Breaking it over a few lines makes it a bit more readable

 ```
        projections = [item 
            for sublist in projections  # Loop 1
            for item in sublist         # Loop 2
        ]
 ```
   
      Or consider using itertools:

 ```
        from itertools import chain
        list(chain.from_iterable(projections))
 ```

  * Is this going to be used to test other code? testutils is a pretty generic name, 
    so perhaps rename it to make this more precise.

## test_enets

  * Is this the only test possible?
    - Make sure to check that the code breaks when you pass in nonsensical input.
    - What happens when num_blocks > 2?
    - If you have multiple tests with a common setup / teardown you could arrange them in a class.
